content/advent-2016/http-client.md

```diff
@@ -0,0 +1,270 @@
+
+## Contexts and Timeouts
+
+Generally your users will give up waiting for an answer after some amount of
+time. As an extreme example, if your checkout flow takes 2 days to return a
+response, your users will probably give up and buy it at Target. 2 days is
+an extreme example, but there is some maximum amount of time your users (and
+your code) should wait for an answer before you should give up and execute the
+fallback logic (like telling users they should try again later).
+
+Complicating this, the user that's calling your library is probably doing so
+as part of a larger request; maybe they are making 10 simultaneous requests to
+your client, and also making a database request and also doing some filesystem
+work. Odds are they want to enforce some deadline for all of that work to
+finish, and if the work isn't complete by then, everything still in progress
+should be canceled.
+
+Go's [`context` library][context] is perfect for this use case. Users can
+create a Context, pass it to multiple threads, and then either cancel the work
+being done in every thread, or time it out after a specific deadline.
+
+[context]: https://golang.org/pkg/context/
+
+Many other languages make it tricky to enforce an absolute deadline on a
+HTTP request - in many languages, you compute the timeout as a duration
+("3 seconds") and that [can reset any time the server sends a single
+byte][byte]. You can use `context.WithDeadline` to [enforce an absolute
+deadline][with-deadline] on a client request, which is really nice.
+
+[byte]: https://kev.inburke.com/slides/reliable-http/#connect-timeout-requests
+[with-deadline]: https://golang.org/pkg/context/#WithDeadline
+
+The best practice is to pass a Context as the first parameter to every function
+that can open a socket. Here are some example function signatures:
+
+```go
+client.Messages.Create(ctx context.Context, to, from string)
+client.Emails.Get(ctx context.Context, sid string)
+client.Notifications.List(ctx context.Context, filters url.Values)
+```
+
+All you have to do inside your library is pass the Context to the http.Request
+object via `request.WithContext()`:
+
+```go
+req, err := http.NewRequest(method, url, body)
+req = req.WithContext(ctx)
+return client.Do(req)
+```
+
+This will let your users coordinate timeouts very precisely, as well as cancel
+requests they no longer need.
+
+## Type Parsing
+
+GRPC API's are becoming more common, but most HTTP API's you'll deal with are
+still returning XML or JSON data. JSON offers only a few types - numbers,
+strings, booleans, and arrays/maps of those. Go offers a much wider range of
+types. Consider trying to marshal those JSON/XML objects into a more useful
+type for your users.
+
+For example, the Twilio API returns phone numbers as strings - `"+14105551234"`
+for example. We can parse those into PhoneNumber objects, and then provide
+helpers to let users print different formats of the number (e.g. `"(410)
```

Comment:
You can also implement `json.Unmarshaler` on that type - which can be useful when you have a key that can be either a string, or an array. 

(which you talk about below... :smile: )